### ISSUE TITLE
InvokeViaAttributeShortcut makes it impossible to pickle objects

### ISSUE THREAD

**Issue by pySilver** [wagtail/wagtail#6408]
### Issue Summary

`wagtail.core.utils.InvokeViaAttributeShortcut` makes it impossible to pickle `BaseSetting` object.

So it effectively kills object level caching.

### Steps to Reproduce

```
import pickle
some_settings_obj = SomeSettings.objects.get(pk=1)
picke.dumps(some_settings_obj, -1)
```
would produce `TypeError: 'str' object is not callable`

**Comment by pySilver** [comment]
I bet the reason for this is slots usage. https://github.com/wagtail/wagtail/blob/5ccc094ea7c37bac4b58518024e764a0ccba91e6/wagtail/core/utils.py#L168

**Comment by pySilver** [comment]
Its not that trivial to fix since `obj` would contain refrence back to the object we are about to pickle, so circular dep. arises. Whats usually is used in this situations is handling on a model level (BaseSettings in our case), by implementing getstate/setstate there that would add/remove field before/after serialization/deserialization.

**Comment by pySilver** [comment]
It would be much easier to use get_page as `@cached_property` so we can filter out this field from serialization like this: https://www.ianlewis.org/en/pickling-objects-cached-properties

**Comment by pySilver** [comment]
Quick dirty fix:

```

    ########
    # Fixes unserializable `page_url` attribute that was added in Wagtail 2.10
    ########
    def __reduce__(self):
        data = self.__getstate__()
        data[DJANGO_VERSION_PICKLE_KEY] = get_version()
        class_id = self._meta.app_label, self._meta.object_name
        return model_unpickle, (class_id,), data

    def __getstate__(self) -> dict:
        """
        Skip cached properties from pickle
        """
        state = self.__dict__.copy()
        del state["page_url"]
        return state

    def __setstate__(self, state):
        state["page_url"] = InvokeViaAttributeShortcut(self, "get_page_url")
        self.__dict__.update(state)
```

**Comment by pySilver** [comment]
If you really want `__slots_ here it is:

```python
class InvokeViaAttributeShortcut:
    """
    Used to create a shortcut that allows an object's named
    single-argument method to be invoked using a simple
    attribute reference syntax. For example, adding the
    following to an object:

    obj.page_url = InvokeViaAttributeShortcut(obj, 'get_page_url')

    Would allow you to invoke get_page_url() like so:

    obj.page_url.terms_and_conditions

    As well as the usual:

    obj.get_page_url('terms_and_conditions')
    """

    __slots__ = "obj", "method_name", "__dict__"

    def __init__(self, obj, method_name):
        self.obj = obj
        self.method_name = method_name

    def __getattr__(self, name):
        method = getattr(self.__dict__["obj"], self.__dict__["method_name"])
        return method(name)

    def __setattr__(self, key, value):
        self.__dict__[key] = value

    def __getstate__(self):
        return self.__dict__

    def __setstate__(self, state):
        self.__dict__.update(state)
```

**Comment by pySilver** [comment]
@ababic ^^

**Comment by ababic** [comment]
@pySilver thanks for letting us know about this, although I'm not sure this fix is quite right. 

When pickling a slotted object, you may find it complains with a misleading TypeError, but this message comes from the oldest protocol, which is the default. You can select the latest protocol with the -1 argument. In Python 2.7 this would be 2 (which was introduced in 2.3), and in 3.6 it is 4.

```
$ pickle.loads(pickle.dumps(f, -1))
<__main__.Foo object at 0x1129C770>
```

in Python 2.7:

```
$ pickle.loads(pickle.dumps(f, 2))
<__main__.Foo object at 0x1129C770>
```

in Python 3.6

```
$ pickle.loads(pickle.dumps(f, 4))
<__main__.Foo object at 0x1129C770>
```

With this in mind, the fix isn't to add those methods to the class, but to ensure whatever is doing the pickling is not doing so with the default pickling protocol, which handles `__slots__` just fine.

**Comment by pySilver** [comment]
@ababic whatis f in your case? Is that a page instance or smth else? I've tested the issue with 3.8 and latest (-1) protocol.

**Comment by ababic** [comment]
@pySilver it's just psueodo code - I don't have my laptop on me. 

If the code you added is indeed required to make slots work, then using slots is not really benefiting us much. Any memory gains would just be cancelled out by the creation of the dict in `__getstate__`. I would rather just drop the `__slots__` in that case.

**Comment by pySilver** [comment]
Yes, dropping slots will do the job. :)

**Comment by ababic** [comment]
@pySilver Unexpectedly, removing `__slots__` does not solve the problem. I have tested, and am seeing the same error regardless of whether `__slots__` is set. It would seem that `InvokeViaAttributeShortcut` just isn't inheriting the `__getstate__` and `__setstate__` default behaviour that the top-level object (the instance of the BaseSetting subclass being pickled) - so it needs those methods adding regardless.

**Comment by ababic** [comment]
@pySilver This should be fixed by #6411, if you care to take a look.

**Comment by gasman** [comment]
Fixed in #6411
